Use LockFile.PackageFolders instead of NuGetPathContext when it is available

In `Microsoft.NET.Build.Tasks.NuGetPackageResolver`, use `LockFile.PackageFolders` instead of `NuGetPathContext` when it is available.

From Justin:

> When resolving packages from a project.lock.json or project.assets.json file the folder paths given in the file should be used instead of NuGetPathContext.
> 
> ResolvePackageDependencies should pass the paths to FallbackPackagePathResolver here: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Packaging/FallbackPackagePathResolver.cs#L26
> 
> Passing NuGetPathContext will re-evaluate the global packages folder path, and that may not match the folder used by restore if it was passed in on the command line to restore. The value in the assets file however will contain exactly what restore used.

When using NuGetPathContext, NuGet finds the paths to the global packages folder and fallback package folders by reading NuGet.Config, including the user wide NuGet.config. This could lead to errors for users running as a service or under a system account which doesn’t have a profile:

> C:\Users\dotnetcoretest2\.nuget\packages\microsoft.net.sdk\1.0.0-alpha-20161029-1\build\Microsoft.PackageDependencyResolution.targets(138,5): error MSB4018: The "ResolvePackageDependencies" task failed unexpectedly. 
> 
> C:\Users\dotnetcoretest2\.nuget\packages\microsoft.net.sdk\1.0.0-alpha-20161029-1\build\Microsoft.PackageDependencyResolution.targets(138,5): error MSB4018: NuGet.Configuration.NuGetConfigurationException: Failed to read NuGet.Config due to unauthorized access. Path: 'C:\WINDOWS\system32\config\systemprofile\AppData\Roaming\NuGet\NuGet.Config'. ---> System.UnauthorizedAccessException: Access to the path 'C:\WINDOWS\system32\config\systemprofile' is denied.

--- a/src/Tasks/Microsoft.NET.Build.Tasks/NuGetPackageResolver.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/NuGetPackageResolver.cs
@@ -2,8 +2,13 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using NuGet.Common;
+using NuGet.Configuration;
 using NuGet.Packaging;
+using NuGet.ProjectModel;
 using NuGet.Versioning;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
 
 namespace Microsoft.NET.Build.Tasks
 {
@@ -16,9 +21,33 @@ namespace Microsoft.NET.Build.Tasks
             _packagePathResolver = new FallbackPackagePathResolver(pathContext);
         }
 
+        public NuGetPackageResolver(string userPackageFolder, IEnumerable<string> fallbackPackageFolders)
+        {
+            _packagePathResolver = new FallbackPackagePathResolver(userPackageFolder, fallbackPackageFolders);
+        }
+
         public string GetPackageDirectory(string packageId, NuGetVersion version)
         {
             return _packagePathResolver.GetPackageDirectory(packageId, version);
+        }
+
+        public static NuGetPackageResolver CreateResolver(LockFile lockFile, string projectPath)
+        {
+            NuGetPackageResolver packageResolver;
+
+            string userPackageFolder = lockFile.PackageFolders.FirstOrDefault()?.Path;
+            if (userPackageFolder != null)
+            {
+                var fallBackFolders = lockFile.PackageFolders.Skip(1).Select(f => f.Path);
+                packageResolver = new NuGetPackageResolver(userPackageFolder, fallBackFolders);
+            }
+            else
+            {
+                NuGetPathContext nugetPathContext = NuGetPathContext.Create(Path.GetDirectoryName(projectPath));
+                packageResolver = new NuGetPackageResolver(nugetPathContext);
+            }
+
+            return packageResolver;
         }
     }
 }

--- a/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageDependencies.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageDependencies.cs
@@ -131,8 +131,7 @@ namespace Microsoft.NET.Build.Tasks
             {
                 if (_packageResolver == null)
                 {
-                    NuGetPathContext nugetPathContext = NuGetPathContext.Create(Path.GetDirectoryName(ProjectPath));
-                    _packageResolver = new NuGetPackageResolver(nugetPathContext);
+                    _packageResolver = NuGetPackageResolver.CreateResolver(LockFile, ProjectPath);
                 }
 
                 return _packageResolver;

--- a/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePublishAssemblies.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePublishAssemblies.cs
@@ -2,11 +2,8 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System.Collections.Generic;
-using System.IO;
 using Microsoft.Build.Framework;
 using Microsoft.Build.Utilities;
-using NuGet.Configuration;
-using NuGet.Frameworks;
 using NuGet.ProjectModel;
 
 namespace Microsoft.NET.Build.Tasks
@@ -44,9 +41,9 @@ namespace Microsoft.NET.Build.Tasks
 
         protected override void ExecuteCore()
         {
-            LockFile lockFile = new LockFileCache(BuildEngine4).GetLockFile(AssetsFilePath);
-            NuGetPathContext nugetPathContext = NuGetPathContext.Create(Path.GetDirectoryName(ProjectPath));
+            LockFile lockFile = new LockFileCache(BuildEngine4).GetLockFile(AssetsFilePath);            
             IEnumerable<string> privateAssetsPackageIds = PackageReferenceConverter.GetPackageIds(PrivateAssetsPackageReferences);
+            IPackageResolver packageResolver = NuGetPackageResolver.CreateResolver(lockFile, ProjectPath);
 
             ProjectContext projectContext = lockFile.CreateProjectContext(
                 NuGetUtils.ParseFrameworkName(TargetFramework),
@@ -54,7 +51,7 @@ namespace Microsoft.NET.Build.Tasks
                 PlatformLibraryName);
 
             IEnumerable<ResolvedFile> resolvedAssemblies = 
-                new PublishAssembliesResolver(new NuGetPackageResolver(nugetPathContext))
+                new PublishAssembliesResolver(packageResolver)
                     .WithPrivateAssets(privateAssetsPackageIds)
                     .Resolve(projectContext);
 

